pynicotine/plugins/autobahn/__init__.py

```diff
@@ -0,0 +1,248 @@
+from pynicotine.pluginsystem import BasePlugin
+from pynicotine.config import config
+#from pynicotine.userbrowse import UserBrowse
+
+#def enable(plugins):
+#    global PLUGIN
+#    PLUGIN = Plugin(plugins)
+
+
+#def disable(plugins):
+#    global PLUGIN
+#    PLUGIN = None
+
+
+class Plugin(BasePlugin):
+    __name__ = "Autobahn"
+    settings = {
+        'ban_message': 'AUTOMATIC MESSAGE: You are not sharing enough and you were banned. You will be unbanned when you will contribute.',
+        'warn_message': 'AUTOMATIC MESSAGE: Thanks for sharing over 10GB. If you share over 20GB I will remove 2 file limit allowing you to queue as many files as you want',
+        'friend_message': 'AUTOMATIC MESSAGE: Welcome to my buddy list. 2 file limit will no longer apply to you - you will need to retry all files stopped by the limit.\nAUTOMATIC MESSAGE: I would appreciate if you don\'t put more than 1000 files in a queue at a time as it may crash Nicotine+',
+        'ban_min_files': 100,
+        'ban_min_bytes': 10000,
+        'friend_min_bytes': 20000
+    }
+    metasettings = {
+        'ban_message': {
+            'description': 'Banned message',
+            'type': 'textview'},
+        'warn_message': {
+            'description': 'Warning message',
+            'type': 'textview'},
+        'friend_message': {
+            'description': 'Added to list message',
+            'type': 'textview'},
+        'ban_min_files': {
+            'description': 'Minimum number of shared files to avoid a ban',
+            'type': 'integer'},
+        'ban_min_bytes': {
+            'description': 'Minimum total size of shared files to avoid a ban (MB)',
+            'type': 'integer'},
+        'friend_min_bytes': {
+            'description': 'Minimum total size of shared files to add to friend list (MB)',
+            'type': 'integer'}
+    }
+
+    def init(self):
+        self.probed = {}
+        self.retries = {}
+
+    def upload_queued_notification(self, user, virtualfile, realfile):
+        try:
+            self.probed[user]
+        except KeyError:
+            self.probed[user] = 'requesting'
+            self.retries[user] = 1
+            self.frame.np.userbrowse.browse_user(user)
+            self.core.queue.append(slskmessages.GetUserStats(user))
+            self.log('User %s, requesting information...' % user)
+
+    def user_stats_notification(self, user, stats):
+        try:
+            status = self.probed[user]
+        except KeyError:
+            # we did not trigger this notification
+            return
+        if status == 'requesting':
+            #check if user file list has finished loading
+            #if not ready - request UserStats again - it is used as a timer hack to prevent locking UI
+            
+            #status check fails if user browse tab is closed manualy before fully loading
+            try:
+                total_size_text = self.frame.userbrowse.pages[user].share_size_label.get_text()
```

Comment:
We don't use references to `frame` in plugins anymore.

You have to consider headless mode, where there is no GUI.